dockstore-integration-testing/src/test/java/io/dockstore/webservice/metrics/MetricsDataS3ClientIT.java

```diff
@@ -56,137 +74,181 @@
+        tool.setName("dockstore-cgpmap");
+        tool.setRegistryString(Registry.QUAY_IO.getDockerPath());
+        tool.setDefaultVersion("symbolic.v1");
+        tool.setDefaultCWLTestParameterFile("/examples/cgpmap/bamOut/bam_input.json");
+        DockstoreTool registeredTool = containersApi.registerManual(tool);
+        registeredTool = containersApi.refresh(registeredTool.getId());
+        containersApi.publish(registeredTool.getId(), CommonTestUtilities.createOpenAPIPublishRequest(true));
+
+        final String toolId = "quay.io/dockstoretestuser2/dockstore-cgpmap";
+        final String toolVersionId = "symbolic.v1";
+        final String toolExpectedS3KeyPrefixFormat = "tool/quay.io/dockstoretestuser2/dockstore-cgpmap/symbolic.v1/%s"; // This is the prefix without the file name. Format by providing the platform
+        extendedGa4GhApi.executionMetricsPost(executions, platform1, toolId, toolVersionId, null);
+        metricsDataList = verifyMetricsDataList(toolId, toolVersionId, 1);
+        metricsData = verifyMetricsDataInfo(metricsDataList, toolId, toolVersionId, platform1, String.format(toolExpectedS3KeyPrefixFormat, platform1));
+        verifyMetricsDataMetadata(metricsData, ownerUserId, "");
+        verifyMetricsDataContent(metricsData, executions);
+        assertEquals(3, getS3ObjectsFromBucket().size(), "There should be 3 objects, 2 for workflows and 1 for tools");
+    }
+
+    @Test
+    void testSubmitMetricsDataErrors() {
+        // Admin user
+        final ApiClient webClient = getOpenAPIWebClient(USER_2_USERNAME, testingPostgres);
+        final WorkflowsApi workflowApi = new WorkflowsApi(webClient);
+        final ExtendedGa4GhApi extendedGa4GhApi = new ExtendedGa4GhApi(webClient);
+        // Non-admin user
+        final ApiClient otherWebClient = getOpenAPIWebClient(OTHER_USERNAME, testingPostgres);
+        final ExtendedGa4GhApi otherExtendedGa4GhApi = new ExtendedGa4GhApi(otherWebClient);
+
+        String id = "#workflow/github.com/DockstoreTestUser2/dockstore_workflow_cnv";
+        String versionId = "master";
+        String platform = Partner.TERRA.name();
+        String description = "A single execution";
+
+        // Test malformed ID
+        ApiException exception = assertThrows(ApiException.class, () -> extendedGa4GhApi.executionMetricsPost(List.of(), platform, "malformedId", "malformedVersionId", null));
+        assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), exception.getCode());
+
+        // Test ID that doesn't exist
+        exception = assertThrows(ApiException.class, () -> extendedGa4GhApi.executionMetricsPost(List.of(), platform, "github.com/nonexistent/id", "master", null));
+        assertEquals(Response.Status.NOT_FOUND.getStatusCode(), exception.getCode(), "Should not be able to submit metrics for non-existent id");
+        assertTrue(exception.getMessage().contains("Tool not found"), "Should not be able to submit metrics for non-existent id");
```

Comment:
Reference a constant for `"Tool not found"` (helps if we ever tweak error messages)